### ISSUE TITLE
issues of the ios report, any papers of those vulnerabilities?

### ISSUE THREAD

**Issue by sunnycomes** [MobSF/Mobile-Security-Framework-MobSF#433]
You said the followings are vulnerabilities, any references about them? How can I make sure those vulnerabilities are truly vulnerable, any papers to refer?

**banned API**
```
"alloca|gets|memcpy|printf|scanf|sprintf|sscanf|strcat|StrCat|strcpy|" +
"StrCpy|strlen|StrLen|strncat|StrNCat|strncpy|StrNCpy|strtok|swprintf|vsnprintf|"+
"vsprintf|vswprintf|wcscat|wcscpy|wcslen|wcsncat|wcsncpy|wcstok|wmemcpy"
```

**Weak Crypto API(s)**
```
"kCCAlgorithmDES|kCCAlgorithm3DES||kCCAlgorithmRC2|kCCAlgorithmRC4|" +
"kCCOptionECBMode|kCCOptionCBCMode"
```

**Crypto API(s)**
```
"CCKeyDerivationPBKDF|CCCryptorCreate|CCCryptorCreateFromData|" +
"CCCryptorRelease|CCCryptorUpdate|CCCryptorFinal|CCCryptorGetOutputLength|" +
"CCCryptorReset|CCCryptorRef|kCCEncrypt|kCCDecrypt|kCCAlgorithmAES128|" +
"kCCKeySizeAES128|kCCKeySizeAES192|kCCKeySizeAES256|kCCAlgorithmCAST|" +
"SecCertificateGetTypeID|SecIdentityGetTypeID|SecKeyGetTypeID|SecPolicyGetTypeID|" +
"SecTrustGetTypeID|SecCertificateCreateWithData|SecCertificateCreateFromData|" +
"SecCertificateCopyData|SecCertificateAddToKeychain|SecCertificateGetData|" +
"SecCertificateCopySubjectSummary|SecIdentityCopyCertificate|" +
"SecIdentityCopyPrivateKey|SecPKCS12Import|SecKeyGeneratePair|SecKeyEncrypt|" +
"SecKeyDecrypt|SecKeyRawSign|SecKeyRawVerify|SecKeyGetBlockSize|" +
"SecPolicyCopyProperties|SecPolicyCreateBasicX509|SecPolicyCreateSSL|" +
"SecTrustCopyCustomAnchorCertificates|SecTrustCopyExceptions|" +
"SecTrustCopyProperties|SecTrustCopyPolicies|SecTrustCopyPublicKey|" +
"SecTrustCreateWithCertificates|SecTrustEvaluate|SecTrustEvaluateAsync|" +
"SecTrustGetCertificateCount|SecTrustGetCertificateAtIndex|SecTrustGetTrustResult|" +
"SecTrustGetVerifyTime|SecTrustSetAnchorCertificates|" +
"SecTrustSetAnchorCertificatesOnly|SecTrustSetExceptions|SecTrustSetPolicies|" +
"SecTrustSetVerifyDate|SecCertificateRef|" +
"SecIdentityRef|SecKeyRef|SecPolicyRef|SecTrustRef"
```

**Weak HASH API(s)**
```
"CC_MD2_Init|CC_MD2_Update|CC_MD2_Final|CC_MD2|MD2_Init|" +
"MD2_Update|MD2_Final|CC_MD4_Init|CC_MD4_Update|CC_MD4_Final|CC_MD4|MD4_Init|" +
"MD4_Update|MD4_Final|CC_MD5_Init|CC_MD5_Update|CC_MD5_Final|CC_MD5|MD5_Init|" +
"MD5_Update|MD5_Final|MD5Init|MD5Update|MD5Final|CC_SHA1_Init|CC_SHA1_Update|" +
"CC_SHA1_Final|CC_SHA1|SHA1_Init|SHA1_Update|SHA1_Final"
```

**HASH API(s)**
```
"CC_SHA224_Init|CC_SHA224_Update|CC_SHA224_Final|CC_SHA224|" +
"SHA224_Init|SHA224_Update|SHA224_Final|CC_SHA256_Init|CC_SHA256_Update|" +
"CC_SHA256_Final|CC_SHA256|SHA256_Init|SHA256_Update|SHA256_Final|" +
"CC_SHA384_Init|CC_SHA384_Update|CC_SHA384_Final|CC_SHA384|SHA384_Init|" +
"SHA384_Update|SHA384_Final|CC_SHA512_Init|CC_SHA512_Update|CC_SHA512_Final|" +
"CC_SHA512|SHA512_Init|SHA512_Update|SHA512_Final"
```

The above code are found in this file:
https://github.com/MobSF/Mobile-Security-Framework-MobSF/blob/ab7c171b8a9e02e659a0080bdf29dc742edcf261/StaticAnalyzer/views/ios/binary_analysis.py

**Comment by DominikSchlecht** [comment]
Hey @sunnycomes,

these checks are there to check weather these function are referenced somewhere in the binary. It does not neccesarally mean that these are vulns (e.g. using an MD5 to match photos may be ok, but MD5 is not OK for storing passwords. memcpy can also be used in a secure way, but is much harder to implement than functions which do not allow memory corruption by default).

Does this answer your question?

Greetings,
Dominik

**Comment by sunnycomes** [comment]
Thank you for your reply. You are right that the listed hash apis are not that secure, and md5 is week. What about Weak Crypto API(s) and Crypto API(s), any papers or documents about them, what are alternatives for them?

**Comment by DominikSchlecht** [comment]
Hey @sunnycomes,

sry, this would go beyond the scope of an issue, but I can point in an [direction](https://developer.apple.com/library/content/documentation/Security/Conceptual/cryptoservices/GeneralPurposeCrypto/GeneralPurposeCrypto.htmll).

From there on I would advise you to visit trainings etc. to gain a deeper understanding and learn best practices.

I hope I was able to help you.

Closing this issue since not really an issue with MobSF.

Greetings,
Dominik
